[Streams] Math Processor

awahab07 opened the pull request:
## Summary

The **Math Processor** enables arithmetic and logical expressions in Streams processing pipelines. It evaluates expressions per-document, computing derived fields from existing data (e.g., `bytes / duration`, `price * quantity`, `cpu_usage > 90`). Results are written to a target field as either a number or boolean.

The processor uses [kbn-tinymath](https://github.com/elastic/kibana/tree/main/src/platform/packages/private/kbn-tinymath) for expression parsing. TinyMath provides a friendly syntax for mathematical operations and is already used in Kibana's Lens formula editor. User entered expressions are parsed into an AST, then Streamlang validates it against the supported function set, and then transpiled to either ES|QL or Painless targets.

The UI is kept similar to Lens formula editor, but uses a curated set of functions/operators supported by Streamlang. The editor (based on `@kbn/code-editor`) includes custom syntax highlighting (functions, operators, numbers, field references), real-time validation with squiggly error markers, and comprehensive autocomplete. Function autocomplete displays signatures with parameter hints and the function category. Autocompletion for field names is sourced from Stream's simulation context.

https://github.com/user-attachments/assets/01e2989f-3e67-490f-be13-0be3e7567f53

---

## Supported Functions & Operators

The math processor uses **OTTL ([OpenTelemetry Transformation Language](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl))** as the baseline for compatibility, ensuring expressions can be transpiled to future OTTL targets in addition to ES|QL and Painless.

### Supported

| Category | Operators/Functions |
|----------|---------------------|
| **Arithmetic** | `+`, `-`, `*`, `/` |
| **Comparison** | `>`, `>=`, `<`, `<=`, `==` |
| **Comparison (function)** | `eq()`, `neq()`, `gt()`, `gte()`, `lt()`, `lte()` |
| **Logarithmic** | `log()` — natural logarithm (base e) |
| **Grouping** | Parentheses `()` |

### Not Supported (OTTL limitation)

The following TinyMath functions are explicitly rejected with helpful error messages:

- **Rounding**: `round`, `ceil`, `floor`
- **Trigonometric**: `sin`, `cos`, `tan`, `asin`, `acos`, `atan`, `sinh`, `cosh`, `tanh`
- **Other math**: `abs`, `sqrt`, `cbrt`, `pow`, `mod`, `exp`, `signum`, `hypot`, `log_ten`
- **Constants**: `pi()`, `e()`, `tau()`
- **Aggregations**: `mean`, `sum`, `min`, `max`, `count`, etc.

awahab07 committed:
Add math processor with basic ES|QL transpilation

- Define MathProcessor type/schema with expression, to, ignore_missing
- Implement TinyMath AST to ES|QL transpiler for arithmetic operators

awahab07 committed:
Add complete ES|QL math function support with validation

- Add function registry, AST validator, and field extractor in shared/math/
- Extend transpiler with where condition and ignore_missing handling
- Add neq() for not-equal comparisons (not native to TinyMath)
- Add 153 unit tests for all new functionality

awahab07 committed:
Add Painless transpiler for math processor

Implement script processor generation from TinyMath expressions with
support for all registry functions, ignore_missing, and where conditions.

awahab07 committed:
Add math processor E2E tests

- Add comprehensive E2E tests for math processor (ES|QL, Ingest Pipeline, cross-compatibility)
- Fix integer vs decimal literal handling in ES|QL math transpiler
- Add Euler's constant (e) to function registry
- Add peggy loader hook for Playwright tests to load .peggy grammar files
- Add math processor steps to multi_step.spec.ts (response_time_sec, throughput_bytes_per_sec)
- Document incompatibilities between Painless and ES|QL:
  - Division by zero: Painless throws ArithmeticException, ES|QL returns null
  - sqrt(-1): Painless returns 'NaN' string, ES|QL returns null
  - Overflow to Infinity: Different handling between engines

awahab07 committed:
feat(streams): Math Processor UI and flexible Painless field access

awahab07 committed:
Merge remote-tracking branch 'origin/main' into streams-program-500_Streamlang_math_processor

# Conflicts:
#	x-pack/platform/packages/shared/kbn-streamlang/index.ts
#	x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/steps/blocks/action/editor.tsx

awahab07 committed:
Math expression type inference and validation

awahab07 committed:
Remove duplicate error message string.

awahab07 committed:
Error display improvement.

awahab07 committed:
Make math processor accept arguments like `manual_ingest_pipeline` and  `remove_by_prefix`.

awahab07 committed:
In simulation, don't throw Streamlang validation error (as it causes 500 error), rather wrap it into a runtime painless error.

awahab07 committed:
Math syntax coloring

awahab07 committed:
Centralize math expression language definition and improve docs

- Create language_definition.ts as single source of truth for all math functions
- Remove duplicate docs.ts, replace with language_docs.ts that generates from definition
- Simplify MathFunctionDefinition: merge description/docDescription, derive parameters from args
- Rename exports to include 'math': BOOLEAN_RETURNING_MATH_FUNCTION_NAMES, getMathParameterNames
- Add function autocomplete and signature help for math expression editor
- Improve documentation structure: move examples up, better formatting
- Remove 'processor' terminology from shared math language code

awahab07 committed:
Add field name autocompletion for math expression editor

- Add field suggestions from useEnrichmentFieldSuggestions hook
- Pass field suggestions to completion provider via closure (React-friendly pattern)
- Re-register completion provider when field suggestions change
- Show function categories (arithmetic, comparison, etc.) in autocomplete
- Reuse existing Suggestion type from autocomplete_selector
- Fields sorted first, then functions in autocomplete dropdown

awahab07 committed:
Fix math processor E2E tests for updated transpiler behavior

- Update nested field path tests to use flat key access pattern
- Update validation error tests: ingest pipeline reports errors at runtime,
  ES|QL throws during transpilation
- Remove unused 'result' variable from test destructuring
- Apply Prettier formatting

awahab07 committed:
Code cleanup and deduplication

- Fix ES|QL CASE syntax in comments (function-style, not SQL-style)
- Extract shared TinyMath utilities to tinymath_utils.ts
- Add language_definition helpers for deriving function lists
- Remove verbose separator comments

awahab07 committed:
Merge remote-tracking branch 'origin/main' into streams-program-500_Streamlang_math_processor

# Conflicts:
#	x-pack/platform/packages/shared/kbn-streamlang/index.ts
#	x-pack/platform/plugins/shared/streams/server/lib/streams/helpers/validate_stream.ts

elasticmachine commented:
Pinging @elastic/obs-onboarding-team (Team:obs-onboarding)

kibanamachine committed:
Changes from node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/maintenance_window   --include-path /api/agent_builder --update

kibanamachine committed:
Changes from make api-docs

awahab07 committed:
Add math processor action metadata.

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/372056)
* Commit: 7a91d351ecad72eaf9626510b2af954e9bf23fb3

### Failed CI Steps
* [Scout Test Run Builder](https://buildkite.com/elastic/kibana-pull-request/builds/372056#019b0ec3-8b83-497e-bd84-98c0cb088084)
* [Scout Test Run Builder](https://buildkite.com/elastic/kibana-pull-request/builds/372056#019b0eee-72d9-42a7-9d50-f9387ea23d36)

### History
* :broken_heart: [Build #372020](https://buildkite.com/elastic/kibana-pull-request/builds/372020) failed 41fb84deb27b6c0eea5b71d2acfa12eb4db4abd9
* :broken_heart: [Build #372015](https://buildkite.com/elastic/kibana-pull-request/builds/372015) failed 00bc21c14d19fe357c3138404f43e7ab9ee1d98b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372056","number":372056,"commit":"7a91d351ecad72eaf9626510b2af954e9bf23fb3"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372020","number":372020,"commit":"41fb84deb27b6c0eea5b71d2acfa12eb4db4abd9"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372015","number":372015,"commit":"00bc21c14d19fe357c3138404f43e7ab9ee1d98b"}],"number":372056}
buildkite-pr-comment-->

awahab07 committed:
Fix: Remove tinymath_utils from barrel exports to prevent peggy loading in all Playwright tests

awahab07 committed:
Remove redundant comment.

awahab07 committed:
Load Peggy in Scout's root to avoid errors when other Scout tests are run in CI.

awahab07 committed:
Merge remote-tracking branch 'origin/main' into streams-program-500_Streamlang_math_processor

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/372209)
* Commit: b61dd0321b93a72d9fc05aa66c2307d40ed0983f

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/372209#019b116f-f29e-470e-b7f4-fa81a8ad96ed)

### History
* :broken_heart: [Build #372139](https://buildkite.com/elastic/kibana-pull-request/builds/372139) failed 2e169e7e4834dabf6bc457d46fdbb81fc54a93f9
* :broken_heart: [Build #372056](https://buildkite.com/elastic/kibana-pull-request/builds/372056) failed 7a91d351ecad72eaf9626510b2af954e9bf23fb3
* :broken_heart: [Build #372020](https://buildkite.com/elastic/kibana-pull-request/builds/372020) failed 41fb84deb27b6c0eea5b71d2acfa12eb4db4abd9
* :broken_heart: [Build #372015](https://buildkite.com/elastic/kibana-pull-request/builds/372015) failed 00bc21c14d19fe357c3138404f43e7ab9ee1d98b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372209","number":372209,"commit":"b61dd0321b93a72d9fc05aa66c2307d40ed0983f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372139","number":372139,"commit":"2e169e7e4834dabf6bc457d46fdbb81fc54a93f9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372056","number":372056,"commit":"7a91d351ecad72eaf9626510b2af954e9bf23fb3"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372020","number":372020,"commit":"41fb84deb27b6c0eea5b71d2acfa12eb4db4abd9"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372015","number":372015,"commit":"00bc21c14d19fe357c3138404f43e7ab9ee1d98b"}],"number":372209}
buildkite-pr-comment-->

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/372214)
* Commit: f5b01824be827ef571ef22fe202147e871db9445

### Failed CI Steps
* [FTR Configs #14](https://buildkite.com/elastic/kibana-pull-request/builds/372214#019b11a5-803c-45cf-8953-4fd6f290ee17)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/372214#019b11a5-803c-45cf-8953-4fd6f290ee17) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/372214/jobs/019b11a5-803c-45cf-8953-4fd6f290ee17/artifacts/019b11cc-2e38-4dea-b747-4c4e78306b1d) FTR Configs #<span></span>14 / serverless observability UI - Cases and Rules Serverless Observability Cases Cases list row actions Severity to low

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/99edc7da39bb23a70ed8e541ef0e202bcdb9b580) | [after](https://github.com/elastic/kibana/commit/f5b01824be827ef571ef22fe202147e871db9445) | diff |
| --- | --- | --- | --- |
| `datasetQuality` | 1021 | 1082 | +61 |
| `streamsApp` | 1302 | 1389 | +87 |
| total |  |  | +148 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/99edc7da39bb23a70ed8e541ef0e202bcdb9b580) | [after](https://github.com/elastic/kibana/commit/f5b01824be827ef571ef22fe202147e871db9445) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 336 | 350 | +14 |

#### Any counts in public APIs
> Total count of every any typed public API. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats any` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/99edc7da39bb23a70ed8e541ef0e202bcdb9b580) | [after](https://github.com/elastic/kibana/commit/f5b01824be827ef571ef22fe202147e871db9445) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 17 | 18 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/99edc7da39bb23a70ed8e541ef0e202bcdb9b580) | [after](https://github.com/elastic/kibana/commit/f5b01824be827ef571ef22fe202147e871db9445) | diff |
| --- | --- | --- | --- |
| `datasetQuality` | 550.3KB | 589.8KB | +39.5KB |
| `streamsApp` | 1.4MB | 1.4MB | +61.9KB |
| total |  |  | +101.5KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/99edc7da39bb23a70ed8e541ef0e202bcdb9b580) | [after](https://github.com/elastic/kibana/commit/f5b01824be827ef571ef22fe202147e871db9445) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 5 | 8 | +3 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/99edc7da39bb23a70ed8e541ef0e202bcdb9b580) | [after](https://github.com/elastic/kibana/commit/f5b01824be827ef571ef22fe202147e871db9445) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 24.7KB | 24.7KB | +72.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/99edc7da39bb23a70ed8e541ef0e202bcdb9b580) | [after](https://github.com/elastic/kibana/commit/f5b01824be827ef571ef22fe202147e871db9445) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 378 | 410 | +32 |


</details>


### History
* :broken_heart: [Build #372209](https://buildkite.com/elastic/kibana-pull-request/builds/372209) failed b61dd0321b93a72d9fc05aa66c2307d40ed0983f
* :broken_heart: [Build #372139](https://buildkite.com/elastic/kibana-pull-request/builds/372139) failed 2e169e7e4834dabf6bc457d46fdbb81fc54a93f9
* :broken_heart: [Build #372056](https://buildkite.com/elastic/kibana-pull-request/builds/372056) failed 7a91d351ecad72eaf9626510b2af954e9bf23fb3
* :broken_heart: [Build #372020](https://buildkite.com/elastic/kibana-pull-request/builds/372020) failed 41fb84deb27b6c0eea5b71d2acfa12eb4db4abd9
* :broken_heart: [Build #372015](https://buildkite.com/elastic/kibana-pull-request/builds/372015) failed 00bc21c14d19fe357c3138404f43e7ab9ee1d98b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372214","number":372214,"commit":"f5b01824be827ef571ef22fe202147e871db9445"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372209","number":372209,"commit":"b61dd0321b93a72d9fc05aa66c2307d40ed0983f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372139","number":372139,"commit":"2e169e7e4834dabf6bc457d46fdbb81fc54a93f9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372056","number":372056,"commit":"7a91d351ecad72eaf9626510b2af954e9bf23fb3"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372020","number":372020,"commit":"41fb84deb27b6c0eea5b71d2acfa12eb4db4abd9"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372015","number":372015,"commit":"00bc21c14d19fe357c3138404f43e7ab9ee1d98b"}],"number":372214}
buildkite-pr-comment-->

flash1293 commented:
High level notes:

It doesn't complain about string values being passed to it - should we error on this if we know something's a string?
<img width="876" height="410" alt="Screenshot 2025-12-12 at 12 29 47" src="https://github.com/user-attachments/assets/4900f60a-a9e2-49de-b21c-4ceca0e0b482" />

Also, an invalid tinymath expression is not failing in the validation phase, only during simulation:
<img width="586" height="204" alt="Screenshot 2025-12-12 at 12 40 52" src="https://github.com/user-attachments/assets/6f3fdb36-021d-47e4-8455-342b1dbabc44" />

Since the tinymath is a string on the API, this is a bit dangerous, it should be caught by the validation


Not sure about the boolean stuff... @LucaWintergerst what do you think, should we include this? I can see use cases for it, but I'm a bit unsure. Same about the fancier math, it's nice but I'm not sure how important it is and I don't want us to bite it later. This would also simplify the PR a bit.

awahab07 commented:
> It doesn't complain about string values being passed to it - should we error on this if we know something's a string?

It does, if the type is detected, it'll complain.

<img width="400" height="843" alt="image" src="https://github.com/user-attachments/assets/d37e2ca4-3fb5-49df-861b-2c8efb243d01" />


<br><br>
> Also, an invalid tinymath expression is not failing in the validation phase, only during simulation:

This is how YAML mode works I guess (see the video below), @Kerry350 do you have any input on this.

https://github.com/user-attachments/assets/32780261-92fd-487b-8e62-1b53f26e459c

awahab07 committed:
Make processor value validation (e.g. math expression) part of

awahab07 committed:
Align errors badge to the left.

awahab07 committed:
Merge remote-tracking branch 'origin/main' into streams-program-500_Streamlang_math_processor

mdbirnstiehl reviewed (approved):
Just a couple of nits, but otherwise LGTM.

awahab07 committed:
PR feedback: copy updates.

awahab07 committed:
Make validation error title generic.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/372614)
* Commit: 03daa33d52d033a08e4dce2543af9cb5027f2cc6

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/372614#019b139e-de6a-4a77-b5fe-c00fcf5b79ef)

### History
* :yellow_heart: [Build #372214](https://buildkite.com/elastic/kibana-pull-request/builds/372214) was flaky f5b01824be827ef571ef22fe202147e871db9445
* :broken_heart: [Build #372209](https://buildkite.com/elastic/kibana-pull-request/builds/372209) failed b61dd0321b93a72d9fc05aa66c2307d40ed0983f
* :broken_heart: [Build #372139](https://buildkite.com/elastic/kibana-pull-request/builds/372139) failed 2e169e7e4834dabf6bc457d46fdbb81fc54a93f9
* :broken_heart: [Build #372056](https://buildkite.com/elastic/kibana-pull-request/builds/372056) failed 7a91d351ecad72eaf9626510b2af954e9bf23fb3
* :broken_heart: [Build #372020](https://buildkite.com/elastic/kibana-pull-request/builds/372020) failed 41fb84deb27b6c0eea5b71d2acfa12eb4db4abd9
* :broken_heart: [Build #372015](https://buildkite.com/elastic/kibana-pull-request/builds/372015) failed 00bc21c14d19fe357c3138404f43e7ab9ee1d98b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372614","number":372614,"commit":"03daa33d52d033a08e4dce2543af9cb5027f2cc6"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372214","number":372214,"commit":"f5b01824be827ef571ef22fe202147e871db9445"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372209","number":372209,"commit":"b61dd0321b93a72d9fc05aa66c2307d40ed0983f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372139","number":372139,"commit":"2e169e7e4834dabf6bc457d46fdbb81fc54a93f9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372056","number":372056,"commit":"7a91d351ecad72eaf9626510b2af954e9bf23fb3"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372020","number":372020,"commit":"41fb84deb27b6c0eea5b71d2acfa12eb4db4abd9"}],"number":372614}
buildkite-pr-comment-->

awahab07 committed:
Merge remote-tracking branch 'origin/main' into streams-program-500_Streamlang_math_processor

awahab07 committed:
Fix i18n conflict.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/372825)
* Commit: d9fecabca66bb1dae9d68c5e2fad49e36c8de78e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/bf8c75b853e390db94045bdc41c8fc553ac848b9) | [after](https://github.com/elastic/kibana/commit/d9fecabca66bb1dae9d68c5e2fad49e36c8de78e) | diff |
| --- | --- | --- | --- |
| `datasetQuality` | 1030 | 1091 | +61 |
| `streamsApp` | 1304 | 1391 | +87 |
| total |  |  | +148 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/bf8c75b853e390db94045bdc41c8fc553ac848b9) | [after](https://github.com/elastic/kibana/commit/d9fecabca66bb1dae9d68c5e2fad49e36c8de78e) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 336 | 351 | +15 |

#### Any counts in public APIs
> Total count of every any typed public API. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats any` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/bf8c75b853e390db94045bdc41c8fc553ac848b9) | [after](https://github.com/elastic/kibana/commit/d9fecabca66bb1dae9d68c5e2fad49e36c8de78e) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 17 | 18 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/bf8c75b853e390db94045bdc41c8fc553ac848b9) | [after](https://github.com/elastic/kibana/commit/d9fecabca66bb1dae9d68c5e2fad49e36c8de78e) | diff |
| --- | --- | --- | --- |
| `datasetQuality` | 556.1KB | 595.7KB | +39.6KB |
| `streamsApp` | 1.4MB | 1.4MB | +62.7KB |
| total |  |  | +102.3KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/bf8c75b853e390db94045bdc41c8fc553ac848b9) | [after](https://github.com/elastic/kibana/commit/d9fecabca66bb1dae9d68c5e2fad49e36c8de78e) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 5 | 8 | +3 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/bf8c75b853e390db94045bdc41c8fc553ac848b9) | [after](https://github.com/elastic/kibana/commit/d9fecabca66bb1dae9d68c5e2fad49e36c8de78e) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 24.7KB | 24.7KB | +72.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/bf8c75b853e390db94045bdc41c8fc553ac848b9) | [after](https://github.com/elastic/kibana/commit/d9fecabca66bb1dae9d68c5e2fad49e36c8de78e) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 378 | 411 | +33 |


</details>


### History
* :broken_heart: [Build #372614](https://buildkite.com/elastic/kibana-pull-request/builds/372614) failed 03daa33d52d033a08e4dce2543af9cb5027f2cc6
* :yellow_heart: [Build #372214](https://buildkite.com/elastic/kibana-pull-request/builds/372214) was flaky f5b01824be827ef571ef22fe202147e871db9445
* :broken_heart: [Build #372209](https://buildkite.com/elastic/kibana-pull-request/builds/372209) failed b61dd0321b93a72d9fc05aa66c2307d40ed0983f
* :broken_heart: [Build #372139](https://buildkite.com/elastic/kibana-pull-request/builds/372139) failed 2e169e7e4834dabf6bc457d46fdbb81fc54a93f9
* :broken_heart: [Build #372056](https://buildkite.com/elastic/kibana-pull-request/builds/372056) failed 7a91d351ecad72eaf9626510b2af954e9bf23fb3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372825","number":372825,"commit":"d9fecabca66bb1dae9d68c5e2fad49e36c8de78e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372614","number":372614,"commit":"03daa33d52d033a08e4dce2543af9cb5027f2cc6"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372214","number":372214,"commit":"f5b01824be827ef571ef22fe202147e871db9445"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372209","number":372209,"commit":"b61dd0321b93a72d9fc05aa66c2307d40ed0983f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372139","number":372139,"commit":"2e169e7e4834dabf6bc457d46fdbb81fc54a93f9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372056","number":372056,"commit":"7a91d351ecad72eaf9626510b2af954e9bf23fb3"}],"number":372825}
buildkite-pr-comment-->

awahab07 commented on src/platform/packages/shared/kbn-scout/src/playwright/index.ts:
```diff
@@ -7,6 +7,9 @@
  */
 
+// Needed for Scout tests dependent on .peggy grammar files (`@kbn/tinymath`)
+import './peggy_setup';
```
Thanks @dmlemeshko for looking into this and creating the PR. In the PR, Scout it still dependent on `@kbn/streams-schema`, which is dependent on `@kbn/streamlang`.

I've tried using the change from your PR [here](https://github.com/elastic/kibana/pull/246285) but get the same [error](https://buildkite.com/elastic/kibana-pull-request/builds/372842#019b1ebe-d853-4045-b447-c8d2c68ab0a4/237-242) on CI.

Major issue is, `@kbn/scout` is dependent on `@kbn/synthtrace`, which is dependent on `@kbn/streamlang`. This is described in [this issue](https://github.com/elastic/kibana/issues/236731). I think this should be holistically solved and the solution PR could address https://github.com/elastic/kibana/issues/236731. 

But could we please do that after the feature freeze and let this PR move forward? This PR makes `@kbn/tinymath` a hard dependency in `@kbn/streamlang`, so skipping tests or even removing tests is not an option here. Since in the current dependency chain, `@kbn/scout` is dependent on `@kbn/streamlang`, all Playwright tests commands referencing `**/playwright.config.ts` would fail as long as Streamlang imports `@kbn/tinymath`.

dmlemeshko reviewed (approved):
`kbn/scout` changes LGTM, we will address dependency issue later.

awahab07 committed:
Merge remote-tracking branch 'origin/main' into streams-program-500_Streamlang_math_processor

awahab07 committed:
Restrict math processor to OTTL-compatible functions and operators

- Remove 26 unsupported functions: sqrt, cbrt, ceil, floor, round, exp,
  signum, sin, cos, tan, asin, acos, atan, atan_two, sinh, cosh, tanh,
  log_ten, hypot, pow (2-arg log and 2-arg round), pi, e, tau
- Keep only OTTL-compatible capabilities: arithmetic operators (+, -, *, /),
  abs(), single-argument log() (natural logarithm), and comparison functions
  (eq, neq, gt, gte, lt, lte)
- Move removed functions to REJECTED_FUNCTIONS with helpful error messages
- Update language_definition.ts to remove unsupported categories
- Simplify transpilers by removing special handling for 2-arg log and round
- Update all unit and E2E tests to reflect the restricted function set
- Fix isConstant reference in completion provider (constants were removed)
- Relax floating-point precision in cross-compatibility log test

awahab07 committed:
Compact rejection/error tests using it.each

- ast_validator.test.ts: Consolidate rejected function tests into single it.each
- function_registry.test.ts: Consolidate isSupportedFunction and isRejectedFunction tests
- math.test.ts (ES|QL): Consolidate validation error tests for rejected functions
- math_processor.test.ts (Ingest): Consolidate error-throwing script tests

awahab07 committed:
Compact E2E rejection tests into single parameterized tests

awahab07 committed:
Fix math editor placeholder to use supported expression

awahab07 committed:
Merge remote-tracking branch 'origin/main' into streams-program-500_Streamlang_math_processor

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/373392)
* Commit: a31ab062316e9564f216ae2b97d043fc88099e70

### Failed CI Steps
* [Jest Tests #9](https://buildkite.com/elastic/kibana-pull-request/builds/373392#019b234a-5b86-4afe-9baf-c13b519ad3fe)
* [Jest Tests #9](https://buildkite.com/elastic/kibana-pull-request/builds/373392#019b2358-bca0-4713-9287-9618c33e93b6)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/373392#019b234a-5b86-4afe-9baf-c13b519ad3fe) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/373392/jobs/019b234a-5b86-4afe-9baf-c13b519ad3fe/artifacts/019b2358-aa9e-4946-b719-c75b009faf2e) Jest Tests #<span></span>9 / validateStreamlang processor value validation should pass validation for valid math expression
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/373392#019b2358-bca0-4713-9287-9618c33e93b6) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/373392/jobs/019b2358-bca0-4713-9287-9618c33e93b6/artifacts/019b2367-07ce-43fc-8046-a0392f05f190) Jest Tests #<span></span>9 / validateStreamlang processor value validation should pass validation for valid math expression

### History
* :green_heart: [Build #372825](https://buildkite.com/elastic/kibana-pull-request/builds/372825) succeeded d9fecabca66bb1dae9d68c5e2fad49e36c8de78e
* :broken_heart: [Build #372614](https://buildkite.com/elastic/kibana-pull-request/builds/372614) failed 03daa33d52d033a08e4dce2543af9cb5027f2cc6
* :yellow_heart: [Build #372214](https://buildkite.com/elastic/kibana-pull-request/builds/372214) was flaky f5b01824be827ef571ef22fe202147e871db9445
* :broken_heart: [Build #372209](https://buildkite.com/elastic/kibana-pull-request/builds/372209) failed b61dd0321b93a72d9fc05aa66c2307d40ed0983f
* :broken_heart: [Build #372139](https://buildkite.com/elastic/kibana-pull-request/builds/372139) failed 2e169e7e4834dabf6bc457d46fdbb81fc54a93f9
* :broken_heart: [Build #372056](https://buildkite.com/elastic/kibana-pull-request/builds/372056) failed 7a91d351ecad72eaf9626510b2af954e9bf23fb3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/373392","number":373392,"commit":"a31ab062316e9564f216ae2b97d043fc88099e70"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372825","number":372825,"commit":"d9fecabca66bb1dae9d68c5e2fad49e36c8de78e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372614","number":372614,"commit":"03daa33d52d033a08e4dce2543af9cb5027f2cc6"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372214","number":372214,"commit":"f5b01824be827ef571ef22fe202147e871db9445"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372209","number":372209,"commit":"b61dd0321b93a72d9fc05aa66c2307d40ed0983f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372139","number":372139,"commit":"2e169e7e4834dabf6bc457d46fdbb81fc54a93f9"}],"number":373392}
buildkite-pr-comment-->

awahab07 committed:
Fix validate_streamlang test to use supported math functions

awahab07 commented:
As a result of the discussion, the PR (both code and description) has been updated to to consider OTTL as the baseline for supported math functions and operators (unlike [ES|QL math functions](https://www.elastic.co/docs/reference/query-languages/esql/functions-operators/math-functions) as initially planned).

Kerry350 commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/steps/blocks/action/math/math_expression_editor.tsx:
```diff
@@ -0,0 +1,295 @@
+              <EuiToolTip
+                content={
+                  isWordWrapped
+                    ? i18n.translate('xpack.streams.math.editorDisableWordWrap', {
```
`xpack.streams.math.editorDisableWordWrap` and `xpack.streams.math.editorEnableWordWrap` could be moved to variables as they're both repeated in this file.

Kerry350 commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/steps/blocks/action/math/math_expression_editor.tsx:
```diff
@@ -0,0 +1,295 @@
+      monaco.editor.setModelMarkers(model, 'math-validation', markers);
+    } else {
+      // Clear markers when valid
+      monaco.editor.setModelMarkers(model, 'math-validation', []);
```
You might want to also add a cleanup function for when the component unmounts, I'm not sure if it really matters but just incase Monaco tries to do something with them in the lifecycle:

```ts
return () => {
    if (modelRef.current) {
      monaco.editor.setModelMarkers(modelRef.current, 'math-validation', []);
    }
  };
```

awahab07 replied:
Yes, it should clear in useEffect's cleanup. Thanks for spotting this.

Will do as a follow up to not risk running the CI cycle on FF day. 

Currently it clears the markers on initialization. Since 'math-validation' model is only used for for this component, it works but its's not future proof.

Kerry350 reviewed (approved):
Couple of minor comments, but really nice work 👏 

We can possibly bring some of the Monaco provider capabilities to the `streamlang-yaml-editor` package, but that can happen later.

Errors reflect the simplification of the functions etc. 

<img width="562" height="101" alt="Screenshot 2025-12-16 at 10 23 25" src="https://github.com/user-attachments/assets/45bc2e06-dc0e-4b75-8fef-61d09bd46e48" />
<img width="640" height="144" alt="Screenshot 2025-12-16 at 10 23 39" src="https://github.com/user-attachments/assets/693c8900-1656-4cd2-b117-537b518df87a" />
<img width="677" height="248" alt="Screenshot 2025-12-16 at 10 24 20" src="https://github.com/user-attachments/assets/82ad8d6c-655f-4fc4-bf82-b59538205d06" />
